File: apps/launcher/internal/launcher/useradd.go
package launcher

import (
	"fmt"
	"os"
	"strings"
)

// useraddHint renders the "run this next" command a successful start prints.
// Every such line is BUILT here rather than typed at the call site, because
// the typed ones drifted: both summaries advertised `--admin` for as long as
// that flag had been gone (roles went with the issuer), so a fresh install's
// first instruction was a command useradd refuses. A flag that does not exist
// cannot be spelled in one place any more.
func useraddHint(repo string) string {
	cmd := "semiont useradd"
	if repo != "" {
		cmd += " --repo " + repo
	}
	return cmd + " --email <email>"
}

const useraddUsage = `Usage: semiont useradd --email <email> [--generate-password] [options]

Create or update a user in a Semiont stack, local or codespace. The ISSUER holds
the account, the profile and the password; this decides which realm is meant and
speaks to it. A local realm is administered from here; a codespace's by its own
launcher, over ssh, so that machine's admin credential never crosses the wire.

The password is never typed as an argument. Creating a user prompts for one on
a terminal, or reads it from stdin when piped:

  semiont useradd --email admin@example.com               # prompts
  cat pw | semiont useradd --email bot@example.com        # scripted

Options:

  --email <email>       User email address (required)
  --generate-password   Generate a random 16-char password (printed once)
  --inactive            Disable the account at the identity provider
  --active              Re-enable a disabled account
  --update              Update an existing user
  --upsert              Create if absent, succeed silently if present
  --password-stdin      Set the password (implied when creating; say it
                        explicitly with --update to CHANGE a password)

There are no role flags. No Semiont route grants access on the basis of a role,
so there is nothing here to grant. There is no display name either: the realm
requires a first and last name, and asks the person for their own at first
sign-in rather than have an administrator guess how to split one string.

Launcher-owned (consumed here, not forwarded):

  --repo <owner/name>   Target that codespace stack
  --runtime <name>      Target the LOCAL stack (selector only, as in stop)
  --help, -h            Show this help

Needs a started stack: the realm is reached at the issuer this KB configures,
and semiont start is what records which config that is. With more than one
stack recorded, the working directory disambiguates (the clone whose local stack
is running means local; a clone whose origin names a codespace stack, with
no local stack, means that one) — anywhere less certain, useradd refuses to
guess: say which with --repo or --runtime.

NOTHING auto-creates an account — local and codespace alike. A fresh realm has
no users at all, so this is how the first one comes to exist, and how every
later user and password change happens.

Examples:
  # First user after a fresh local start (prompts for the password)
  semiont useradd --email admin@example.com

  # A second user on a codespace KB
  semiont useradd --repo The-AI-Alliance/my-kb --email alice@example.com --generate-password
`

// Useradd implements `semiont useradd`. The ISSUER owns the account, the
// profile and the password hashing; this decides which realm is meant and
// speaks to it.
//
// A LOCAL stack is administered here, through the same admin client
// `identity sync` uses — no container, and the gateway need not be running.
// A CODESPACE runs its OWN launcher over ssh, for the same reason: the remote
// realm's admin password lives in that machine's environment and never crosses
// the wire.
//
// The password NEVER travels in argv. It used to ride into the container as an
// env var (readable via `inspect` for the stack's whole lifetime); then as an
// exec argument, redacted in the echo — but redaction is cosmetic: `ps` shows
// any process's command line to every user on the host, and the caller's shell
// wrote it to history besides. Now the launcher reads it (prompting on a
// terminal, else from stdin) and pipes it to `--password-stdin`, so it exists
// only in two process memories and the pipe between them.
func Useradd(args []string) int {
	u := NewUI(false)
	for _, a := range args {
		if a == "--help" || a == "-h" {
			fmt.Print(useraddUsage)
			return 0
		}
	}
	if len(args) == 0 {
		fmt.Print(useraddUsage)
		return 1
	}

	// Every flag is READ here — the local path acts on them directly — and
	// every flag but --repo and --runtime is also FORWARDED, because the
	// codespace path still hands them to the gateway's bin. --repo and
	// --runtime select a stack and never cross.
	repo, wantLocal := "", false
	generate, update, wantStdin := false, false, false
	o := useraddOpts{}
	rest := make([]string, 0, len(args))
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--email":
			if i+1 >= len(args) {
				u.Fail("Missing value for --email")
				return 1
			}
			o.email = args[i+1]
		case "--upsert":
			o.upsert = true
		case "--active":
			o.active = true
		case "--inactive":
			o.inactive = true
		case "--repo":
			if i+1 >= len(args) {
				u.Fail("Missing value for --repo")
				return 1
			}
			repo = args[i+1]
			i++
			continue
		case "--runtime": // selector only, mirroring stop: "the local stack"
			if i+1 >= len(args) {
				u.Fail("Missing value for --runtime")
				return 1
			}
			wantLocal = true
			i++
			continue
		case "--password":
			// Removed, not deprecated. It put the secret in argv — visible in
			// `ps` on the host and in the container, kept by the runtime's
			// container record, and written to the caller's shell history.
			u.Fail("--password is no longer accepted: a password in argv is visible to every process on the host.")
			fmt.Fprintln(os.Stderr, "  Let it prompt:   semiont useradd --email <email>")
			fmt.Fprintln(os.Stderr, "  Or pipe it:      cat pw | semiont useradd --email <email>")
			fmt.Fprintln(os.Stderr, "  Or generate it:  semiont useradd --email <email> --generate-password")
			return 1
		case "--generate-password":
			generate = true
			o.generate = true
		case "--update":
			update = true
			o.update = true
		case "--password-stdin":
			wantStdin = true
			o.stdin = true
			continue // re-added below, exactly once
		default:
			// An unknown flag is REFUSED, not forwarded. The far end used to
			// refuse it; now that this is the far end for a local stack,
			// silently ignoring one would let `--admin` — which several
			// documents still advertise and which grants nothing — look like
			// it worked.
			if strings.HasPrefix(args[i], "--") {
				u.Fail("Unknown flag: %s", args[i])
				fmt.Fprintln(os.Stderr, "  See:  semiont useradd --help")
				return 1
			}
		}
		rest = append(rest, args[i])
	}

	// Asking for both a password and a generated one is a contradiction, and
	// it must be REFUSED here: --password-stdin is stripped above and re-added
	// only when a password is actually read, so forwarding alone would let the
	// gateway's own mutual-exclusion check never see the pair — the user would
	// silently get a generated password they did not ask to keep.
	if generate && wantStdin {
		u.Fail("--password-stdin and --generate-password are contradictory: one supplies a password, the other invents one.")
		return 1
	}
	if !useraddValidate(u, o) {
		return 1
	}

	// Which stack? The shared knowledge-verb ladder (stackselect.go).
	target, ok := SelectVerbStack(u, "useradd", LoadStackSet(), repo, wantLocal)
	if !ok {
		return 1
	}

	// The password is read LAST, after every refusal this command can make.
	// Nobody should be asked to type a secret by an invocation that was
	// already going to be rejected for contradictory flags or a missing
	// gateway — the prompt would also bury the actual error.
	//
	// Who supplies it? The gateway requires one only to CREATE, so an --update
	// that isn't explicitly changing the password needs none, and
	// --generate-password means the gateway invents its own.
	password := ""
	if !generate && (!update || wantStdin) {
		pw, ok := readPassword(u)
		if !ok {
			return 1
		}
		password = pw
		rest = append(rest, "--password-stdin")
	}

	if target != nil {
		return useraddCodespace(u, target, rest, password)
	}
	return useraddLocal(u, o, password)
}

// useraddValidate: the mutual exclusions and the one format check. They belong
// here rather than at the far end because a refusal the caller can make is one
// the caller should make — and on the codespace path the far end is an ssh hop
// away.
func useraddValidate(u *UI, o useraddOpts) bool {
	if o.email == "" {
		u.Fail("--email is required")
		return false
	}
	if !strings.Contains(o.email, "@") || strings.ContainsAny(o.email, " \t") ||
		!strings.Contains(o.email[strings.Index(o.email, "@"):], ".") {
		u.Fail("invalid email format: %s", o.email)
		return false
	}
	if o.update && o.upsert {
		u.Fail("--update and --upsert are contradictory: one demands the account exist, the other tolerates it.")
		return false
	}
	if o.inactive && o.active {
		u.Fail("--inactive and --active are contradictory.")
		return false
	}
	return true
}

// useraddCodespace runs the same verb one hop further out: ssh into the
// codespace and run ITS launcher, which administers that stack's realm exactly
// as this one administers a local stack.
//
// The realm's admin password never crosses the wire. It lives in the
// codespace's own environment, and the launcher over there reads it from there
// — this machine neither holds it nor needs to.
//
// CRITICAL: `gh codespace ssh -- cmd` runs the remote side through a SHELL
// (proven live — a `/workspaces/*` glob expands there, which is how the KB
// root is reached). So every argument is single-quote escaped before it
// crosses. The password is exempt by never being an argument: it goes down
// ssh's stdin to `--password-stdin`.
func useraddCodespace(u *UI, st *StackState, args []string, password string) int {
	if !requireGh(u, "useradd against a codespace stack") {
		return 1
	}
	// Build the remote command ONCE, and echo that same string — the
	// launcher's echoed lines are meant to be the exact command it runs (the
	// same contract --dry-run keeps). Echoing the pre-quoting args instead
	// would print something that behaves differently if pasted: $VARs would
	// expand and values with spaces would split.
	remote := remoteUseraddCmd(args, password != "")
	sshArgs := []string{"codespace", "ssh", "-c", st.Codespace.Name, "--", remote}
	u.Log("useradd on %s %s", u.Bold(st.Codespace.Repo), u.Dim("(codespace "+st.Codespace.Name+")"))
	u.EchoCmd("gh", "codespace", "ssh", "-c", st.Codespace.Name, "--", remote)
	if err := runVisibleWithStdin(password, "gh", sshArgs...); err != nil {
		u.Fail("useradd failed inside the codespace (see output above).")
		fmt.Fprintln(os.Stderr, "  Is the stack up?  semiont status --repo "+st.Codespace.Repo)
		return 1
	}
	return 0
}

// remoteUseraddCmd composes the command the codespace's shell will run: cd to
// the KB clone, then the launcher's own useradd. The glob is what the remote
// shell is for — a codespace mounts the repo under /workspaces/<name>, and the
// name is the repo's, not ours to guess.
//
// Nothing here carries a secret, so nothing needs redacting: the password goes
// by stdin, and the realm admin's password is already in that machine's
// environment.
func remoteUseraddCmd(args []string, _ bool) string {
	cmd := "cd /workspaces/* && semiont useradd"
	for _, a := range args {
		cmd += " " + shellQuote(a)
	}
	return cmd
}

// shellQuote wraps a value for a POSIX shell: single quotes protect
// everything except a single quote itself, which is closed, escaped, and
// reopened. Nothing inside can be interpreted as shell syntax.
func shellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", `'\''`) + "'"
}

File: apps/gateway/src/cli/useradd.ts
/**
 * Create or update an account at this knowledge base's identity provider.
 * Invoked inside the gateway container by `semiont useradd`, which execs it and
 * passes every flag through verbatim.
 *
 * (No shebang — tsup's `banner` adds one to every entry.)
 *
 * A user is ONE thing now: the account at the issuer, which owns the credential,
 * mints the `sub` its tokens carry, and decides whether the person may sign in
 * (`--active` / `--inactive` set that there). Semiont keeps no row of its own —
 * everything the gateway knows about a caller it reads off the token, and the
 * identity is a DID derived from those claims rather than a local id.
 *
 * So this command no longer touches a database, and there is no display name or
 * role to pre-create: the issuer holds the profile, and nothing here grants
 * access on the basis of a role.
 */

import * as crypto from 'crypto';
import { loadEnvironmentConfig } from '@semiont/core/node';
import { KeycloakAdminApi } from '../identity/keycloak-admin';

interface Options {
  email: string;
  passwordStdin: boolean;
  generatePassword: boolean;
  active: boolean;
  inactive: boolean;
  update: boolean;
  upsert: boolean;
}

const USAGE = `Usage: semiont-useradd --email <email> [--password-stdin | --generate-password] [options]

Creates the account at this knowledge base's identity provider. Creating a user
requires a password, so one of --password-stdin or --generate-password. Updating
an existing one does not: pass --password-stdin only when the point is to CHANGE
the password.

Semiont stores nothing of its own about a user — the issuer holds the account and
the profile, and the gateway reads what it needs off the token. There are no role
flags, because no route here grants access on the basis of a role, and no display
name: the realm asks the person for their own at first sign-in.

  --email <email>       User email address (required)
  --password-stdin      Read the password from stdin, first line (min 8 chars)
  --generate-password   Generate a random password (printed once)
  --inactive            Disable the account at the identity provider
  --active              Re-enable a disabled account
  --update              Update an existing user
  --upsert              Create if absent, succeed silently if present
  --help, -h            Show this help
`;

function parseArgs(argv: string[]): Options {
  const o: Options = {
    email: '', passwordStdin: false, generatePassword: false,
    inactive: false, active: false, update: false, upsert: false,
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const value = (): string => {
      const v = argv[i + 1];
      if (v === undefined || v.startsWith('--')) throw new Error(`Missing value for ${a}`);
      i++;
      return v;
    };
    switch (a) {
      case '--email': o.email = value(); break;
      case '--password-stdin': o.passwordStdin = true; break;
      case '--generate-password': o.generatePassword = true; break;
      case '--inactive': o.inactive = true; break;
      case '--active': o.active = true; break;
      case '--update': o.update = true; break;
      case '--upsert': o.upsert = true; break;
      case '--help': case '-h': process.stdout.write(USAGE); process.exit(0);
      default: throw new Error(`Unknown flag: ${a}`);
    }
  }
  return o;
}

/**
 * Read the password from stdin's FIRST LINE.
 *
 * A password must never travel in argv: `ps` shows a process's command line to
 * every other user on the host, `docker inspect`/`container inspect` keep it as
 * long as the container record lives, and the caller's shell records it in
 * history. Stdin has none of those properties.
 */
async function readPasswordFromStdin(): Promise<string> {
  // Stop at the first newline rather than draining to EOF: a password is one
  // line, and waiting for the stream to close would hang an interactive run
  // (`docker exec -it … --password-stdin`) after the user pressed Enter, until
  // they thought to send Ctrl-D.
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    const buf = Buffer.from(chunk);
    chunks.push(buf);
    if (buf.includes(0x0a)) break;
  }
  // split() on empty input yields [''], but noUncheckedIndexedAccess types the
  // index access as possibly-undefined regardless — empty stdin is a refusal
  // either way, two lines down.
  const first = Buffer.concat(chunks).toString('utf8').split('\n', 1)[0] ?? '';
  const password = first.replace(/\r$/, '');
  if (!password) throw new Error('--password-stdin was given but stdin carried no password');
  if (password.length < 8) throw new Error('Password must be at least 8 characters long');
  return password;
}

/** Same shape the old CLI produced: 16 base64 chars from 12 random bytes. */
function generatePassword(): string {
  return crypto.randomBytes(12).toString('base64');
}

function validate(o: Options): void {
  if (!o.email) throw new Error('--email is required');
  if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(o.email)) {
    throw new Error(`invalid email format: ${o.email}`);
  }
  if (o.passwordStdin && o.generatePassword) {
    throw new Error('--password-stdin and --generate-password are mutually exclusive');
  }
  if (o.update && o.upsert) {
    throw new Error('--update and --upsert are mutually exclusive');
  }
  if (o.inactive && o.active) {
    throw new Error('--inactive and --active are mutually exclusive');
  }
}

/**
 * The issuer this knowledge base trusts, and the administrator credentials for
 * it — refusing, with the reason, wherever the answer is not this command's to
 * give. An `oidc` issuer is somebody else's directory: Semiont has no standing
 * to create accounts in it, and pretending otherwise would fail at the API
 * rather than here, where the operator can read why.
 */
function keycloakTarget(identity: { type: string; issuer: string } | undefined): {
  issuer: string; username: string; password: string;
} {
  if (!identity) {
    throw new Error(
      'This knowledge base configures no identity provider — add an [identity] section before creating users.',
    );
  }
  if (identity.type !== 'keycloak') {
    throw new Error(
      `Identity type "${identity.type}" is an issuer Semiont does not administer. ` +
        `Create the account at ${identity.issuer}, then it can sign in here.`,
    );
  }
  const username = process.env.KC_BOOTSTRAP_ADMIN_USERNAME;
  const password = process.env.KC_BOOTSTRAP_ADMIN_PASSWORD;
  if (!username || !password) {
    throw new Error(
      'KC_BOOTSTRAP_ADMIN_USERNAME and KC_BOOTSTRAP_ADMIN_PASSWORD must both be set to administer the realm.',
    );
  }
  return { issuer: identity.issuer, username, password };
}

async function main(argv: string[]): Promise<number> {
  const o = parseArgs(argv);
  validate(o);

  // `null`: this bin is exec'd INSIDE the gateway container
  // (`container exec semiont-gateway semiont-useradd`), which mounts no
  // knowledge base and sets no SEMIONT_ROOT. The staged ~/.semiontconfig is
  // the config, and the loader reads it either way.
  const config = loadEnvironmentConfig(null);
  const target = keycloakTarget(config.services.identity);
  const keycloak = await KeycloakAdminApi.connect(target.issuer, target.username, target.password);

  const account = await keycloak.findUserByEmail(o.email);

  let password: string | undefined;
  if (o.generatePassword) {
    password = generatePassword();
    // Printed once and never stored: the caller's only chance to capture it.
    process.stdout.write(`Generated password: ${password}\n`);
  } else if (o.passwordStdin) {
    password = await readPasswordFromStdin();
  } else if (!account) {
    throw new Error('Password required: use --password-stdin or --generate-password');
  }

  if (account && o.upsert) {
    process.stdout.write(`User already exists: ${o.email}\n`);
    return 0;
  }
  if (account && !o.update) {
    throw new Error(
      `User ${o.email} already exists. Use --update to modify or --upsert to skip silently.`,
    );
  }
  if (!account && o.update) {
    throw new Error(`User ${o.email} not found. Remove --update to create a new user.`);
  }

  // Whether the person may sign in is the issuer's to hold, so it is written
  // there and nowhere else. Absent both flags this command does not touch the
  // account's state.
  //
  // Nothing here sets a display name. The realm requires `firstName` and
  // `lastName`, so Keycloak collects them from the person at first sign-in —
  // see keycloakUserProfile in the launcher's identity.go for why that is the
  // person's job and not an administrator's.
  let subject: string;
  if (account) {
    if (password) await keycloak.setPassword(account.id, password);
    if (o.inactive || o.active) await keycloak.setEnabled(account.id, o.active);
    subject = account.id;
  } else {
    subject = await keycloak.createUser(o.email, password!, !o.inactive);
  }

  process.stdout.write(`${account ? 'User updated' : 'User created'}: ${o.email}\n`);
  process.stdout.write(`  Subject: ${subject}\n`);
  if (o.inactive) process.stdout.write('  Disabled at the identity provider\n');
  if (o.active) process.stdout.write('  Enabled at the identity provider\n');
  return 0;
}

main(process.argv.slice(2))
  .then((code) => process.exit(code))
  .catch((error) => {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exit(1);
  });
